# 4.5: Data Consistency Checks

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Importing  “products.csv” data set from your “Original Data"
df_prods = pd.read_csv(r'C:\Users\peter\Documents\04-2024 Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [4]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [8]:
# Importing orders_wrangled
df_ords = pd.read_csv(r'C:\Users\peter\Documents\04-2024 Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [10]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [11]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# Mixed-Type Data

In [ ]:
# Create a dataframe
df_test = pd.DataFrame()

In [13]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [14]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [19]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

C:\Users\peter\AppData\Local\Temp\ipykernel_9904\772068622.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [20]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [21]:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [22]:
# Finding missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [26]:
df_prods.shape

(49693, 5)

In [27]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [28]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [29]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [30]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [31]:
df_prods_clean.shape

(49677, 5)

In [32]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
df_prods_clean_no_dups.shape

(49672, 5)

# Tidying Up and Exporting Changes

In [35]:
# use the string shortcut PATH
path = r'C:\Users\peter\Documents\04-2024 Instacart Basket Analysis'

In [37]:
#Exporting df_prods_clean_no_dups dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK


In [38]:
#task 2
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# ANSWER:

There are many missing values in the 'days_since_prior_order' column, as its count is lower than the other columns´.

'order_number' has min value as 1 and max value as 100. Q2 or median is 11. No discrepancies found.

'days_since_prior_order' has min value as 0 and max as 30. Q2 or median is 7. No discrepancies found.

In [43]:
#task 3
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# No mixed type of data has been found

In [44]:
# There's no mixed data present in the df_ords dataframe
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [45]:
# Task 4 - nothing to do as no mixed type of data found


In [46]:
# Task 5 Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# The 'days_since_prior_order' column is missing 206209 values
# An explanation for this quanitty of missing values could be customers' first orders don't have a value, but I would say this is relevant information that should be preserved instead of removing

In [47]:
# Task 6 Address the missing values using an appropriate method.
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [48]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [49]:
# Fill NAN with 7 (median)
df_ords['days_since_prior_order'].fillna(7, inplace=True)

In [50]:
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     7.0  
11                        2                 11                     7.0  
26                        1                 14                     7.0  
39                        6                 11                     7.0  
45                        3                 12                     7.0  
...                     ...                ...                     ...  
3420930                   4                 12                     7.0  
3420934                   3                 18                     7.0  
3421002                   6                 19                     7.0  
3421019                   1                 15                     7.0  
3421069                   3                 11                     7.0  

[206209 rows x 8 columns]

In [51]:
#TAsk 7
#Run a check for duplicate values in your df_ords data.

df_ords_dups = df_ords[df_ords.duplicated()]

In [52]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# There are no duplicates because it returned empty. If there were duplicates I would execute '.drop_duplicates()' to creat a new datafreame without duplicates.

In [53]:
df_ords_dups.shape

(0, 8)

# TAsk 9 Exporting dataframes

In [ ]:
#Exporting df_prods_clean_no_dups dataframe done in line 37


In [54]:
#Exporting df_ords 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))